File: src/PMSMCurrentRef.c
/**
 * PMSMCurrentRef.c
 * 
 *  Created on: Nov 22 2021
 *      Author: Liam Flaherty
 */

#include "PMSMCurrentRef.h"

#include <math.h>
#include "constants.h"
#include "sat.h"


void PMSMCurrentRefInit(void)
{
    // empty
}

void PMSMCurrentRefStep(const PMSMCurrentRef_Params_t* params,
                        const PMSMCurrentRef_Input_t* in,
                        PMSMCurrentRef_Output_t* out)
{
    // Calculate torque limits
    float wMech = in->we / params->polePairs; 
    float speed = fmaxf(fabsf(wMech), params->Pmax / params->Tmax);
    out->tqLim = (in->Vdc / params->Vnom) * fminf(params->Tmax, params->Pmax / speed);
    out->tqRefSat = sat(in->tqRef, -(out->tqLim), out->tqLim);

    // params needed for current ref generation
    float we_rpm = in->we * (30.0f / M_PI) / params->polePairs;
    float back_emf = params->Ke * we_rpm;
    float VphMax = in->Vdc * ONE_SQRT3;
    if (VphMax == 0.0f) {
        VphMax = 0.1f;
    }

    // voltage modulation index
    float M = back_emf / VphMax;

    float iq;
    float id;

    float I_nom = 2.0f * (out->tqRefSat) / (3.0f * params->polePairs * params->fluxLink);
    if (M <= params->Modulation_Index_Threshold) {
        // zero d-axis control (ZDAC)
        id = 0.0f;
        iq = sat(I_nom, -(params->Imax), params->Imax);
    } else {
        // Field weakening
        id = -params->Imax *
            (M - params->Modulation_Index_Threshold) /
            (params->Modulation_Index_FwMax - params->Modulation_Index_Threshold);
        id = sat(id, -params->Imax, 0);

        float iq_lim = sqrtf(powf(params->Imax, 2.0f) - powf(id, 2.0f));
        iq = sat(I_nom, -iq_lim, iq_lim);
    }

    out->idqRef = (idq_T){
        .id = id,
        .iq = iq,
    };
}

File: sim/src/PMSMCurrentRef.m
classdef PMSMCurrentRef < matlab.System & matlab.system.mixin.Propagates
    % Implements field weakening of idqRef input

    % Public, tunable properties
    properties
        Vnom = 0;
        Pmax = 0;
        Tmax = 0;
        Imax = 0;
        p = 1;
        psim = 0;
        Ld = 0;
        Ke = 0;
        Modulation_Index_Threshold = 0;
        Modulation_Index_FwMax = 0;
    end

    properties(DiscreteState)

    end

    % Pre-computed constants
    properties(Access = private)

    end

    methods(Access = protected)
        function setupImpl(obj)
            % Perform one-time calculations, such as computing constants
        end

        function [idqRef, TqRefSat, TqLim] = stepImpl(obj, TqRef, we, Vdc)
            % Calculate torque limits
            wMech = we / obj.p;
            TqLim = (Vdc/obj.Vnom) * min(obj.Tmax, obj.Pmax / sat(abs(wMech), obj.Pmax/obj.Tmax, inf));
            TqRefSat = sat(TqRef, -TqLim, TqLim);

            % params needed for current ref generation
            we_rpm = we * (30 / pi) / obj.p;
            back_emf = obj.Ke * we_rpm;
            VphMax = Vdc / sqrt(3);
            if VphMax == 0
                VphMax = 0.1;
            end

            % Modulation index
            M = back_emf / VphMax;

            I_nom = 2 * TqRefSat / (3 * obj.p * obj.psim);
            if M <= obj.Modulation_Index_Threshold
                % Zero d axis control
                id = 0;
                iq = sat(I_nom, -obj.Imax, obj.Imax);
            else
                % Field weakening
                id = ...
                    -obj.Imax * ...
                    (M - obj.Modulation_Index_Threshold) / ...
                    (obj.Modulation_Index_FwMax - obj.Modulation_Index_Threshold);
                id = sat(id, -obj.Imax, 0);

                iq_lim = sqrt(obj.Imax^2 - id^2);
                iq = sat(I_nom, -iq_lim, iq_lim);
            end
            
            idqRef = [id; iq];
        end
        
        function resetImpl(obj)
            % Initialize / reset discrete-state properties
        end
    
        %% Output sizing
        function [sz1, sz2, sz3] = getOutputSizeImpl(~)            
            sz1 = [2,1];
            sz2 = 1;
            sz3 = 1;
        end
        
        function [fz1, fz2, fz3] = isOutputFixedSizeImpl(~)
          %Both outputs are always fixed-sized
          fz1 = true;
          fz2 = true;
          fz3 = true;
        end
        
        function [dt1, dt2, dt3] = getOutputDataTypeImpl(~)
            dt1 = 'double';
            dt2 = 'double';
            dt3 = 'double';
        end
    
        function [cp1, cp2, cp3] = isOutputComplexImpl(~)
            cp1 = false;
            cp2 = false;
            cp3 = false;
        end
    end
end
